### ISSUE TITLE
[Mailer] TLS wrongfully enabled

### ISSUE THREAD

**Issue by dbrekelmans** [symfony/symfony#45077]
### Symfony version(s) affected

4.4 and higher

### Description

Given we're using the Mailer component to send an email
When our host is not equal to `localhost`
And we use `smtp` as the scheme in the dsn string
Then TLS should be disabled

In reality, TLS is unexpectedly enabled.

This is caused by a combination of:

```php
// mailer/Transport/Smtp/EsmtpTransportFactory.php

$tls = 'smtps' === $dsn->getScheme() ? true : null;
```

```php
// mailer/Transport/Smtp/EsmtpTransport.php

if (null === $tls) {
    if (465 === $port) {
        $tls = true;
    } else {
        $tls = \defined('OPENSSL_VERSION_NUMBER') && 0 === $port && 'localhost' !== $host;
    }
}
```

### How to reproduce

Set the mailer dsn to `smtp://` with a host other than `localhost`, for example: `smtp://docker-service-name-1`.
Observe that symfony resolves the scheme to `smtps` causing TLS to be enabled

### Possible Solution

PR: https://github.com/symfony/symfony/pull/45076

**Comment by fabpot** [comment]
This is the expected behavior. when using `smtp`, TLS is determined based on the port.

**Comment by bizurkur** [comment]
I have a similar issue where I need TLS to be forcibly disabled and not do these extra checks to enable it automatically. The only workaround I've found so far is to create a custom transport factory:

```php

namespace Acme;

use Symfony\Component\Mailer\Transport\AbstractTransportFactory;
use Symfony\Component\Mailer\Transport\Dsn;
use Symfony\Component\Mailer\Transport\Smtp\SmtpTransport;
use Symfony\Component\Mailer\Transport\TransportInterface;

class SmtpTransportFactory extends AbstractTransportFactory
{
    public function create(Dsn $dsn): TransportInterface
    {
        $transport = new SmtpTransport(null, $this->dispatcher, $this->logger);

        $stream = $transport->getStream();
        $stream->setHost($dsn->getHost());
        $stream->setPort($dsn->getPort(0));
        $stream->disableTLS();

        return $transport;
    }

    protected function getSupportedSchemes(): array
    {
        return ['smtp'];
    }
}
```

And then override the default `mailer.transport_factory.smtp` service
```xml
        <service id="mailer.transport_factory.smtp" class="Acme\SmtpTransportFactory" parent="mailer.transport_factory.abstract">
            <tag name="mailer.transport_factory" priority="-100" />
        </service>
```
----------

My use case: In the process of switching from Swift Mailer to Symfony Mailer. This is in a development environment. We use a local mail catcher tool to catch all emails. When the default EsmtpTransport sets up the socket to (by default) include TLS settings and test the connection it's crashing our mail catcher tool.

Personally, I think it would be awesome if there were an option to forcibly disable any TLS magic when I know TLS should not be used, e.g. `smtp://127.0.0.1:1025?tls=false`. It's cool when things can auto-detect settings, but being able to manually control those settings is also cool.

**Comment by snpy** [comment]
@dbrekelmans 
@fabpot 
I've just encountered "issue" at hand where my DSN was actually explicitly setting the 25 port (`smpt://internal.domain:25`).

Even with this port set explicitly the capabilities check:

```
$capabilities = $this->getCapabilities($response);
```

kicks in and 3rd criteria in:

```
if (!$stream->isTLS() && \defined('OPENSSL_VERSION_NUMBER') && \array_key_exists('STARTTLS', $capabilities)) {
```

forces TLS to start; the port itself actually does not do anything, actually it may be causing the issue.

@fabpot 
I think the problem with `SocketStream::isTls` is that it's being used differently in two different contexts:

1. In `EsmtpTransport`'s c-tor we're using it to determine if stream should use TLS or not.
2. In `EsmtpTransport::doHeloCommand` we're using it to "check" if TLS was started.

In c-tor we can clearly see that TLS will be disabled (assuming the `$tls` argument is set to NULL) when any of:

* port is not 465 nor 0 (setting it to anything else will cause `$tls` to be set to FALSE,
* OPENSSL_VERSION_NUMBER is not defined,
* host is "localhost"

is met.

So looking at those conditions whenever DSN string is set to "smtp://anything.but.localhost:25" we're ending up with stream.tls = FALSE.

However it seems the `doHeloCommand` method treats the `stream.tls` differently - not as in "do not utilise TLS", but instead "make sure TLS is setup" (when required).

This time the criteria for TLS being used are not the same as in c-tor; now we'll want to have TLS when:

* stream.tls is FALSE (it was not yet setup),
* OPENSSL_VERSION_NUMBER is defined,
* SMTP server's capabilities have `STARTTLS` in them.

In my case I do have a clear SMTP server misconfiguration (the server should not respond with "STARTTLS" when command EHLO is run).
However the EstmpTransport class seems to be misusing the `SocketStream::$tls` in either c-tor or `doHeloCommand` method.

Sorry if I'm missing anything; just wanted to share an hour of debugging in case someone finds in useful.

To wrap things up; in my case instead of using this:

```
smpt://internal.domain:25
```

I had to use this:

```
smpt://internal.domain:25?verify_peer=false
```

to stop getting `SSL routines:tls_process_server_certificate:certificate verify failed` errors, which are there, because we're utilising non-public domain form SMTP.
(this is not the fix for about, but rather a workaround; still using TLS on port 25, but at least not getting SSL exceptions anymore)
